### PR TITLE
feat(trace): audit-trail observability — events, query, stats, OCSF export, replay, bounded memory (#175, #176, #177, #179, #182, #213)

### LINKED ISSUE #175: Record handle expansions and policy denials in the audit trail
## Summary

Make `Kernel.expand()` and denied grant attempts produce first-class audit records in
the `TraceStore`, and populate real provenance on expansion Frames (currently
`principal_id=""`).

## Why this matters

I-02 says every authorized execution must be auditable. Handle expansion returns
additional rows of a stored dataset — a data-access event with real exfiltration
relevance — yet it only emits a log line. Policy denials are likewise only logged,
so `explain()` cannot answer "who was refused what, when, and why", which is half
the value of an authorization audit trail.

## Current evidence

- `Kernel.expand` (in `kernel/__init__.py`) performs constraint rechecks and logging but writes nothing to `kernel.traces`.
- `handles.py:315`: expansion Frames carry `Provenance(principal_id="", action_id=action_id)` — the principal is known at expand time but not recorded.
- `PolicyDenied` paths in `grant_capability` raise and log without a trace entry.
- `trace.py` `TraceStore` has no record type for non-invoke events.

## External context

The OWASP Agent Observability Standard models denials and data-access events as
auditable events, not just successful actions (https://aos.owasp.org/spec/trace/extend_ocsf/).

## Proposed implementation

1. Decide the record shape: either extend `ActionTrace` with an `event_type`
   (`invoke | expand | deny`) or add sibling dataclasses in `trace.py`.
2. Record expansions with principal, handle id, capability, page/fields requested,
   and outcome; record denials with principal, capability, and stable reason code
   from `policy_reasons.py`.
3. Fill expansion `Provenance.principal_id` from the verified token.
4. Surface both through `explain()`/trace listing so #147 (CLI audit) and #126
   (persistence) inherit them.

## AI-agent execution notes

- Inspect first: `trace.py`, `kernel/__init__.py` (`expand`, `grant_capability`), `handles.py` (`expand`), `policy_reasons.py`, `tests/test_trace.py`, `tests/test_handles.py`.
- Keep `ActionTrace` backward compatible (additive fields with defaults) — adapters and OTel export read it.
- Edge cases: expansion of an expired handle; denial during `request_capabilities` vs `grant`.

## Acceptance criteria

- A successful `expand()` produces a retrievable audit record naming the principal and handle.
- A `PolicyDenied` grant produces an audit record with the stable reason code.
- Expansion Frames carry a non-empty `principal_id` in provenance.
- Existing trace tests pass unchanged.

## Test plan

Extend `tests/test_trace.py` and `tests/test_handles.py`; assert records and provenance.
Run `make ci`.

## Documentation plan

Update `docs/architecture.md` (audit pipeline) and `docs/security.md` (audit guarantees);
CHANGELOG `Added`.

## Migration and compatibility notes

Additive fields with defaults; not expected to require migration. Note the new record
types for downstream trace consumers.

## Risks and tradeoffs

More trace volume (bounded once ISSUE 10 lands); schema growth in `trace.py` must stay
within the 300-line module budget — split if needed.

## Suggested labels

reliability, security, architecture

### PR DESCRIPTION
**Pull request by dgenio**
## What changed

One cohesive change set across the **TraceStore / audit-trail observability** area (everything anchored on `TraceStore`, `ActionTrace`, and the kernel recording path), closing the recommended triage group.

- **`models.py`** — `ActionTrace` gains additive `event_type` (`invoke`/`expand`/`deny`) and `reason_code` (defaults preserve the original `invoke` meaning).
- **`trace.py` / new `trace_query.py`** — `TraceQuery` + pure `query_traces()` (filter by principal, capability, event type, outcome, reason code, `since`-inclusive/`until`-exclusive window; deterministic `(invoked_at, action_id)` order + pagination). `TraceStore` is now bounded (`max_entries`, oldest-first eviction, loud first eviction, `evicted_count`).
- **`stores/_protocols.py`, `sqlite.py`, `jsonl.py`, `memory.py`, `_trace_codec.py`** — `query()` added to `TraceStoreProtocol` and all trace backends; revocation backends track each token's `expires_at` and `sweep_expired()` (never un-revoking a live token). **Breaking:** `RevocationStoreProtocol.track()` now takes `expires_at`.
- **`tokens.py`** — threads `expires_at` into `track`; adds `HMACTokenProvider.sweep_revocations()`.
- **new `stats.py`** — `KernelStats` collector + immutable `StatsSnapshot`; wired at kernel choke points (grant, invoke, fallback, firewall warnings, downgrade, handle store/expand). Exposed as `Kernel.stats` / `Kernel.reset_stats()`.
- **new `ocsf.py`** — `trace_to_ocsf()` / `traces_to_ocsf()` map any record to OCSF API Activity (6003) events, AOS-enriched; pure, dependency-free.
- **new `replay.py`** — `DecisionRecord`, `record_decision()`, `replay() -> DecisionDiff` (allow→deny / deny→allow / reason-code flips; rate-limit flips surfaced separately).
- **`kernel/__init__.py` + new `kernel/_audit.py`** — record `deny` traces on `PolicyDenied` and `expand` traces on `Kernel.expand()`; `Kernel.query_traces()`.
- **`handles.py`** — expansion Frames now carry `Provenance.principal_id`.
- **Docs/examples** — `architecture.md`, `security.md`, `integrations.md` (OCSF mapping table), `capabilities.md` (replay), `trace_export.md`, `CHANGELOG.md`; runnable offline `examples/ocsf_export_demo.py` and `examples/trace_replay_demo.py` (added to `make example` and CI).

Closes #175, #176, #177, #179, #182, #213.

## Why

These six issues share one code area, data model, and implementation path — what/how-much the audit subsystem records (#175, #182), how it is read (#177, #179), and how it is exported/replayed (#176, #213) — so they are cleanest as one change. Developed in **Mode B** per the requester: new latitude allowed, retro-compat not required (hence the `track()` signature change), one combined PR.

## How verified

`make ci` (`ruff format --check` → `ruff check` → `mypy src/` → `pytest -q --cov` → all examples):

- `ruff format --check` — 107 files already formatted
- `ruff check src/ tests/ examples/` — All checks passed
- `mypy src/` — Success: no issues found in 57 source files
- `pytest -q` — **715 passed, 1 skipped** (new suites: `test_trace_query.py`, `test_stats.py`, `test_ocsf.py`, `test_replay.py`; extended `test_trace.py`, `test_tokens.py`, `test_handles.py`, `test_kernel.py`, `test_stores_sqlite.py`)
- `make ci` exit code `0`; both new examples run clean offline.

## Tradeoffs / risks

- **Breaking change (intentional, Mode B):** `RevocationStoreProtocol.track()` adds an `expires_at` parameter and the protocol gains `sweep_expired()` — custom revocation backends must update. In-tree backends and tests are updated.
- **OCSF schema validation** is structural (field/type assertions + golden file) and offline, not against a vendored AOS schema (AOS is young; the mapping is versioned and isolated in `weaver_kernel.ocsf`).
- **Replay fidelity** validates policy *structure*, not argument-dependent rules (trace args are redacted); rate-limit-dependent flips are surfaced separately because they are replay-order-sensitive.
- A token *revoked after its expiry entry was already swept* leaves a benign lingering `_revoked` entry (the token is expired and fails the verifier's expiry check regardless); normal lifecycle (revoke while live → sweep after expiry) is fully bounded.

## Scope notes

Limited to the recommended triage group. Adjacent observability items intentionally left as follow-ups: wiring `KernelStats` into `otel.py` as gauges (#179 notes this is acceptable as a follow-up), and `#125` (OTel/LangWatch `ActionTrace` export), which partially overlaps the already-shipped `instrument_kernel` + `export_action_traces`.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_019446VfpRWBaPqU4WX5KgTX

---
_Generated by [Claude Code](https://claude.ai/code/session_019446VfpRWBaPqU4WX5KgTX)_

### CODE REVIEW

File: src/weaver_kernel/trace.py [hunk 1]
```diff
@@ -74,6 +78,8 @@ def export_action_trace(
         "handle_id": trace.handle_id,
         "sensitivity": trace.sensitivity.value,
         "status": "failed" if trace.error is not None else "succeeded",
+        "event_type": trace.event_type,
+        "reason_code": trace.reason_code,
         "error": trace.error,
```
[the remark this answers is not in the document]
Comment (dgenio) [reply src/weaver_kernel/trace.py:85]: Good catch — this is addressed in commit `2e7dbbf`. The `export_action_trace()` docstring now documents `event_type` (`invoke`/`expand`/`deny`) and `reason_code`, and explicitly states that a denial *does* appear in the export as a `deny` event with `status == "failed"` (see the `Returns:` section). It no longer claims denials are never exported.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/weaver_kernel/trace_query.py [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+        jointly complete.
+
+    Raises:
+        AgentKernelError: If ``offset`` or ``limit`` is negative.
+    """
+    if query.offset < 0:
+        raise AgentKernelError(f"TraceQuery.offset must be >= 0, got {query.offset}.")
+    if query.limit is not None and query.limit < 0:
+        raise AgentKernelError(f"TraceQuery.limit must be >= 0 or None, got {query.limit}.")
+
```
[the remark this answers is not in the document]
Comment (dgenio) [reply src/weaver_kernel/trace_query.py:109]: Addressed in commit `2e7dbbf`. `trace_query.py` now has an `_as_utc()` helper that treats a naive `since`/`until` as UTC, and `_matches()` wraps both bounds with it before comparing against the always-aware `invoked_at`, so naive datetimes no longer raise `TypeError`. This mirrors the UTC-normalization used elsewhere in the codebase.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/weaver_kernel/kernel/_stream.py [hunk 3]
```diff
@@ -155,6 +156,12 @@ async def invoke_stream_impl(
                 error=None if yielded_any else "stream produced no chunks",
             )
         )
+        kernel._stats.on_invocation(
+            failed=not yielded_any,
+            fallback=False,
+            redacted=bool(last_frame.warnings) if last_frame else False,
+            downgraded=initial_mode != response_mode,
+        )
```
[the remark this answers is not in the document]
Comment (dgenio) [reply src/weaver_kernel/kernel/_stream.py:169]: Addressed in commit `2e7dbbf`. `_stream.py` now accumulates `redacted_any = redacted_any or bool(frame.warnings)` across *every* yielded frame (not just `last_frame`) and passes that into `on_invocation(redacted=...)`, so per-chunk redaction warnings on earlier frames are no longer under-counted.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/weaver_kernel/stores/memory.py [hunk 4]
```diff
@@ -33,10 +47,14 @@ def revoke(self, token_id: str) -> None:
...
 
-    def track(self, principal_id: str, token_id: str) -> None:
+    def track(self, principal_id: str, token_id: str, expires_at: datetime.datetime) -> None:
         """Record that *token_id* was issued to *principal_id*."""
         with self._lock:
             self._principal_tokens.setdefault(principal_id, set()).add(token_id)
+            self._expiry[token_id] = expires_at
+            self._track_count += 1
+            if self._track_count % _SWEEP_INTERVAL == 0:
+                self._sweep_locked(datetime.datetime.now(tz=datetime.timezone.utc))
```
[the remark this answers is not in the document]
Comment (dgenio) [reply src/weaver_kernel/stores/memory.py:64]: Addressed in commit `2e7dbbf`. `InMemoryRevocationStore.track()` now normalizes a naive `expires_at` to UTC (`replace(tzinfo=timezone.utc)`) before storing, consistent with `SQLiteRevocationStore`, which prevents the `TypeError` in `_sweep_locked()`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/weaver_kernel/stores/memory.py [hunk 5]
```diff
@@ -51,3 +69,29 @@ def revoke_principal(self, principal_id: str) -> int:
...
             return len(newly_revoked)
+
+    def sweep_expired(self, now: datetime.datetime) -> int:
+        """Drop revocation/tracking state for tokens expired at *now*.
+
+        Returns:
+            The number of tracked tokens whose state was removed.
+        """
+        with self._lock:
+            return self._sweep_locked(now)
```
[the remark this answers is not in the document]
Comment (dgenio) [reply src/weaver_kernel/stores/memory.py:92]: Addressed in commit `2e7dbbf`. `sweep_expired()` now normalizes a naive `now` to UTC before delegating to `_sweep_locked()`, so a caller passing naive-as-UTC `now` no longer triggers a `TypeError` against stored aware expiries.

---
_Generated by [Claude Code](https://claude.ai/code)_
